core/src/main/java/feast/core/job/dataflow/DataflowJobManager.java

```diff
@@ -298,4 +323,75 @@ private String waitForJobToRun(DataflowPipelineJob pipelineResult)
       Thread.sleep(2000);
     }
   }
+
+  /**
+   * GetJobLabels returns a key, value map of labels that should be applied to the DataflowJob.
+   *
+   * <p>A Dataflow job can ingest data from multiple FeatureSets. Each FeatureSet can have its own
+   * key, value labels. When multiple FeatureSets have the same label key, the label value for the
+   * DataflowJob will be the concatenation of all the FeatureSets label values with the same key.
+   *
+   * <p>The label key and value for the Dataflow job will follow the requirements defined here:
+   * https://cloud.google.com/resource-manager/docs/creating-managing-labels#requirements. If the
+   * label does not fulfill the requirements, the label will be ignored rather than errors being
+   * thrown. This is because invalid labels should not result in failed Dataflow jobs and disruption
+   * to the ingestion of feature data.
+   *
+   * <p>Warning logs from Feast Core can be inspected to debug cases where expected labels for
+   * Dataflow jobs are not applied.
+   *
+   * @param featureSets FeatureSets that the DataflowJob expects to ingest from
+   * @return Labels that should be applied to the Dataflow job
+   */
+  private Map<String, String> getJobLabels(Collection<FeatureSet> featureSets) {
```

Comment:
1. Currently we're decoupling jobs from core and job doesn't have access to models anymore. At best it should be FeatureSetProto
2. JobManager knows nothing about FeatureSets as for now, and it's not his responsibility. I suggest to move this label calculation to JobCoordinator. (but it will only have FeatureSetProto as well)

Reply:
1. @pyalex Do you mean this field and method on the Job class are going to be removed? 
https://github.com/feast-dev/feast/blob/master/core/src/main/java/feast/core/model/Job.java#L88
https://github.com/feast-dev/feast/blob/master/core/src/main/java/feast/core/model/Job.java#L110

> By the way, I have run end-to-end test manually with this update and I can confirm the labels get propagated from the FeatureSets to the Dataflow jobs. So currently the the FeatureSets information can be retrieved from the Job.

2. I agree that the job label calculation can be moved to JobCoordinator. However, I believe this makes more sense if we have made either changes:
- add labels field to `IngestionJob.proto` so when the JobManager submit the job it knows what labels to apply
OR
- update JobManager interface with labels argument i.e  
 from `Job startJob(Job job)` to maybe  `Job startJob(Job job, Map<String,String> labels)`

However, I don't feel like making changes to the interface and Job proto, so that they are more stable and because we don't really have a clear use case yet of adding labels to Job proto aside from being used in Dataflow jobs.

So currently, I made the job labels application Dataflow job specific to avoid unnecessary changes and also because the label requirements for Dataflow job will be different than for other kinds of Job (if we have more in the future). Other kind of jobs may not even support labels at all. 

Do you have any thoughts on this @woop? Whether we should update the JobManager interface or IngestionJob proto.

Reply:
There's actually bigger problem that I forgot to mention. Job's featureSets will change during job lifetime without restarting/updating job since we delivery featureSets through kafka right now in-flight.
I'm not sure we want to restart job just to update it's labels.
As alternative proposal: I currently work on separating JobManagement from core, thus we will be able to run many JM services and each can be configured to use it's own set of labels. Hence, we will have groups of feature sets labeled.

Reply:
1. Yes I meant those fields. Please check this [PR](https://github.com/feast-dev/feast/pull/903) however it's still under review. But in general we have plan to completely separate JobManagement & Core models. So there won't be only one direction dependency JM -> Core. Thus Core's models don't know about jobs.

Reply:
2. We can add labels to the Job model itself (calculated on JobCoordinator level). And each JobManager can decide what to do with them.